packages/@aws-cdk/aws-servicecatalog/lib/private/association-manager.ts

```diff
@@ -0,0 +1,90 @@
+import * as cdk from '@aws-cdk/core';
+import { TagUpdateConstraintOptions } from '../constraints';
+import { IPortfolio } from '../portfolio';
+import { IProduct } from '../product';
+import { CfnPortfolioProductAssociation, CfnResourceUpdateConstraint } from '../servicecatalog.generated';
+import { hashValues } from './util';
+import { InputValidator } from './validation';
+
+/**
+ * Simplify type for constraint map to hold the L1 constraint and association resources.
+ */
+type AssociationOrConstraint = CfnPortfolioProductAssociation | CfnResourceUpdateConstraint;
+
+export class AssociationManager {
+  public static associateProductWithPortfolio(scope: cdk.Resource, portfolio: IPortfolio, product: IProduct): void {
+    const associationKey = this.getAssociationKey(scope, portfolio, product);
+    if (this.checkConstraintOrAssocationExists(associationKey, Constraints.ASSOCIATION)) {
+      const association = new CfnPortfolioProductAssociation(scope, `PortfolioProductAssociation${associationKey}`, {
+        portfolioId: portfolio.portfolioId,
+        productId: product.productId,
+      });
+
+      const associationValue = new Map<Constraints, AssociationOrConstraint>();
+      associationValue.set(Constraints.ASSOCIATION, association);
+      this.associationMap.set(associationKey, associationValue);
+    }
+  }
+
+  public static allowTagUpdates(
+    scope: cdk.Resource, portfolio: IPortfolio, product: IProduct,
+    options: TagUpdateConstraintOptions,
+  ): void {
+    InputValidator.validateLength(this.prettyPrintAssociation(portfolio, product), 'description', 0, 2000, options.description);
+    const associationKey = this.ensureProductIsAssociatedWithPortfolio(scope, portfolio, product);
+    if (this.checkConstraintOrAssocationExists(associationKey, Constraints.RESOURCE_UPDATE)) {
+      const constraint = new CfnResourceUpdateConstraint(scope, `ResourceUpdateConstraint${associationKey}`, {
+        acceptLanguage: options.messageLanguage,
+        description: options.description,
+        portfolioId: portfolio.portfolioId,
+        productId: product.productId,
+        tagUpdateOnProvisionedProduct: options.allowUpdatingProvisionedProductTags === false ? 'NOT_ALLOWED' : 'ALLOWED',
+      });
+
+      // Add dependsOn to force proper order in deployment.
+      constraint.addDependsOn(this.associationMap.get(associationKey)!.get(Constraints.ASSOCIATION)!);
+      this.associationMap.get(associationKey)!.set(Constraints.RESOURCE_UPDATE, constraint);
+    } else {
+      throw new Error(`Cannot have multiple resource update constraints for association ${this.prettyPrintAssociation(portfolio, product)}`);
+    }
+  }
+
+  private static associationMap = new Map<string, Map<Constraints, AssociationOrConstraint>>();
```

Comment:
So, this design means that there can never be a Constraint that is associated multiple times with the same Product-Portfolio pair.

Is this true? Do all constraints can only be associated once with a given Product-Portfolio pair, like the tag update constraint is?

Reply:
the answer is fuzzy.  For the resource update, launch rule, launch role, and stackset the answer is yes, only one can be associated with a portfolio-product.  For launch notifications it could have multiple, but we had another way of handling that potentially by tracking the background `Topics` that were added.  Regardless, if we get rid of the association map this is no longer defined.